document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java

```diff
@@ -0,0 +1,77 @@
+package uk.gov.companieshouse.document.generator.accounts.mapping.smallfull.mappers;
+
+import uk.gov.companieshouse.accountsdates.AccountsDatesHelper;
+import uk.gov.companieshouse.accountsdates.impl.AccountsDatesHelperImpl;
+import uk.gov.companieshouse.api.model.company.CompanyProfileApi;
+import uk.gov.companieshouse.api.model.company.account.LastAccountsApi;
+import uk.gov.companieshouse.api.model.company.account.NextAccountsApi;
+import uk.gov.companieshouse.document.generator.accounts.mapping.smallfull.model.ixbrl.period.Period;
+
+import java.time.LocalDate;
+
+
+public abstract class ApiToPeriodMapperDecorator implements ApiToPeriodMapper {
+
+    private final ApiToPeriodMapper apiToPeriodMapper;
+
+    private AccountsDatesHelper accountsDatesHelper = new AccountsDatesHelperImpl();
+
+    public ApiToPeriodMapperDecorator(ApiToPeriodMapper apiToPeriodMapper) {
+        this.apiToPeriodMapper = apiToPeriodMapper;
+    }
+
+    @Override
+    public Period apiToPeriod(CompanyProfileApi companyProfile) {
+
+        Period period = apiToPeriodMapper.apiToPeriod(companyProfile);
+
+        boolean isSameYear = isSameYearFiler(companyProfile);
+
+        setCurrentPeriod(period, companyProfile, isSameYear);
+        setPreviousPeriod(period, companyProfile, isSameYear);
+
+        return period;
+    }
+
+    private void setCurrentPeriod(Period period, CompanyProfileApi companyProfile, boolean isSameYear) {
+
+        NextAccountsApi nextAccounts = companyProfile.getAccounts().getNextAccounts();
+        period.setCurrentPeriodStartOnFormatted(convertToDisplayDate(nextAccounts.getPeriodStartOn()));
+        period.setCurrentPeriodEndOnFormatted(convertToDisplayDate(nextAccounts.getPeriodEndOn()));
+        period.setCurrentPeriodBSDate(convertDateToBSDate(nextAccounts.getPeriodStartOn(),
+                nextAccounts.getPeriodEndOn(), isSameYear));
+    }
+
+    private void setPreviousPeriod(Period period, CompanyProfileApi companyProfile, boolean isSameYear) {
+
+        if (isMultipleYearFiler(companyProfile)) {
+            LastAccountsApi lastAccounts = companyProfile.getAccounts().getLastAccounts();
+            period.setPreviousPeriodStartOnFormatted(convertToDisplayDate(lastAccounts.getPeriodStartOn()));
+            period.setPreviousPeriodEndOnFormatted(convertToDisplayDate(lastAccounts.getPeriodEndOn()));
+            period.setPreviousPeriodBSDate(convertDateToBSDate(lastAccounts.getPeriodStartOn(),
+                    lastAccounts.getPeriodEndOn(), isSameYear));
+        }
+    }
+
+    private boolean isMultipleYearFiler(CompanyProfileApi companyProfile) {
+        LastAccountsApi lastAccountsApi = companyProfile.getAccounts().getLastAccounts();
+        return lastAccountsApi != null && lastAccountsApi.getPeriodEndOn() != null;
```

Comment:
Just wondering, we are checking ` lastAccounts.getPeriodEndOn() != null`. 
Wouldn't we need to check `lastAccounts.getPeriodStartOn() != null`? Considering we then passed it to `convertToDisplayDate(lastAccounts.getPeriodStartOn())`.

Reply:
we are checking if it is a multiple year filer. So we check if there is a lastAccountApi and if that has ended with GetPeriodEndOn. Else it is not a multi year file.